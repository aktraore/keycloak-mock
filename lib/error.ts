export class DuplicateUserError extends Error {
    constructor(message: string) {
        super(message);

        Object.setPrototypeOf(this, DuplicateUserError.prototype);

        this.name = 'DuplicateUserError';
    }
}

export class UserNotFoundError extends Error {
    constructor(message: string) {
        super(message);

        Object.setPrototypeOf(this, UserNotFoundError.prototype);

        this.name = 'UserNotFoundError';
    }
}
