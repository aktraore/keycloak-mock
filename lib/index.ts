export { Mock, MockOptions, activateMock, deactivateMock, getMock, getMockInstance } from './mock';
export {
    MockInstance,
    MockInstanceParams,
    CreateMockInstanceOptions,
    createMockInstance,
} from './instance';
export {
    default as MockDatabase,
    MockUser,
    MockUserCredentialType,
    MockUserProfile,
    CreateMockUserOptions,
} from './database';
export { default as createBearerToken, CreateTokenOptions } from './createBearerToken';
export { DuplicateUserError, UserNotFoundError } from './error';
