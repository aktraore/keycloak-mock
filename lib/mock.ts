import qs from 'qs';
import nock, { Scope } from 'nock';

import { NockClientRequest, ViewFn, DeleteViewFn, PostViewFn } from './types';
import { MockInstance } from './instance';
import { decodeTokenAndAttachUser } from './middlewares';
import {
    listCertificates,
    getUser,
    deleteUser,
    getUserInfo,
    createToken,
    createUser,
    listUsers,
    getUmaConfiguration,
    getOpenIdConfiguration,
    validateToken,
} from './views';

let __activeMocks__: Map<string, Mock> = new Map<string, Mock>();

export interface Mock {
    scope: Scope;
    instance: MockInstance;
}

export interface MockOptions {
    listCertificatesView?: ViewFn;
    getUserView?: ViewFn;
    deleteUserView?: DeleteViewFn;
    getUserInfoView?: ViewFn;
    listUsersView?: ViewFn;
    getUmaConfiguration?: ViewFn;
    getOpenIdConfiguration?: ViewFn;
    createTokenView?: PostViewFn;
    createUserView?: PostViewFn;
    validateToken?: PostViewFn;
}

const decodeBody = (request: NockClientRequest, requestBody: unknown): {} => {
    const contentType = (request.headers['content-type'] || '').split(';')[0];

    switch (contentType) {
        case 'application/x-www-form-urlencoded':
            return qs.parse(requestBody as string);

        // JSON is handled by nock already
        default:
            return requestBody as {};
    }
};

const activateMock = (instance: MockInstance, options?: MockOptions): Mock => {
    const { authServerURL, realm, clientID } = instance.params;

    const existingMock = __activeMocks__.get(authServerURL);
    if (existingMock) {
        throw new Error(`There is an existing mock active for ${authServerURL}`);
    }

    const scope = nock(authServerURL)
        .persist()
        .get(`/realms/${realm}/protocol/openid-connect/certs`)
        .reply(async function () {
            await decodeTokenAndAttachUser(instance, this.req);

            if (options && options.listCertificatesView) {
                return options.listCertificatesView(instance, this.req);
            }

            return listCertificates(instance, this.req);
        })
        .get(new RegExp(`/admin/realms/${realm}/users/(.+)`))
        .reply(async function () {
            await decodeTokenAndAttachUser(instance, this.req);

            if (options && options.getUserView) {
                return options.getUserView(instance, this.req);
            }

            return getUser(instance, this.req);
        })
        .delete(new RegExp(`/admin/realms/${realm}/users/(.+)`))
        .reply(async function () {
            await decodeTokenAndAttachUser(instance, this.req);

            if (options && options.deleteUserView) {
                return options.deleteUserView(instance, this.req);
            }

            return deleteUser(instance, this.req);
        })
        .get(`/realms/${realm}/protocol/openid-connect/userinfo`)
        .reply(async function () {
            await decodeTokenAndAttachUser(instance, this.req);

            if (options && options.getUserInfoView) {
                return options.getUserInfoView(instance, this.req);
            }

            return getUserInfo(instance, this.req);
        })
        .get(`/admin/realms/${realm}/users`)
        .query(() => true)
        .reply(async function () {
            await decodeTokenAndAttachUser(instance, this.req);

            if (options && options.listUsersView) {
                return options.listUsersView(instance, this.req);
            }

            return listUsers(instance, this.req);
        })
        .get(`/realms/${realm}/.well-known/uma2-configuration`)
        .reply(async function () {
            await decodeTokenAndAttachUser(instance, this.req);

            if (options && options.getUmaConfiguration) {
                return options.getUmaConfiguration(instance, this.req);
            }

            return getUmaConfiguration(instance, this.req);
        })
        .get(`/realms/${realm}/.well-known/openid-configuration`)
        .reply(async function () {
            await decodeTokenAndAttachUser(instance, this.req);

            if (options && options.getOpenIdConfiguration) {
                return options.getOpenIdConfiguration(instance, this.req);
            }

            return getOpenIdConfiguration(instance, this.req);
        })
        .post(`/realms/${realm}/protocol/openid-connect/token`)
        .reply(async function (uri, body) {
            const decodedBody = decodeBody(this.req, body);

            if (options && options.createTokenView) {
                return options.createTokenView(instance, this.req, decodedBody);
            }

            return createToken(instance, this.req, decodedBody);
        })
        .post(`/realms/${realm}/protocol/openid-connect/token/introspect`)
        .reply(async function (uri, body) {
            const decodedBody = decodeBody(this.req, body);

            await decodeTokenAndAttachUser(instance, this.req);

            if (options && options.validateToken) {
                return options.validateToken(instance, this.req, decodedBody);
            }

            return validateToken(instance, this.req, decodedBody);
        })
        .post(`/admin/realms/${realm}/users`)
        .reply(async function (uri, body) {
            const decodedBody = decodeBody(this.req, body);

            await decodeTokenAndAttachUser(instance, this.req);

            if (options && options.createUserView) {
                return options.createUserView(instance, this.req, decodedBody);
            }

            return createUser(instance, this.req, decodedBody);
        });

    const mock = { scope, instance };
    __activeMocks__.set(authServerURL, mock);

    return mock;
};

const deactivateMock = (mock: Mock): void => {
    const { authServerURL } = mock.instance.params;

    const existingMock = __activeMocks__.get(authServerURL);
    if (!existingMock) {
        throw new Error(`No active mock for ${authServerURL}`);
    }

    __activeMocks__.delete(authServerURL);

    mock.scope.persist(false);

    // @ts-ignore
    mock.scope.keyedInterceptors = {};

    // @ts-ignore
    mock.scope.interceptors = [];
};

const getMock = (authServerURL: string): Mock => {
    const mock = __activeMocks__.get(authServerURL);
    if (!mock) {
        throw new Error(`No active mock for ${authServerURL}`);
    }

    return mock;
};

const getMockInstance = (authServerURL: string): MockInstance => {
    return getMock(authServerURL).instance;
};

export { activateMock, deactivateMock, getMock, getMockInstance };
