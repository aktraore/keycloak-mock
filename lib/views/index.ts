export { default as listCertificates } from './listCertificates';
export { default as getUser } from './getUser';
export { default as deleteUser } from './deleteUser';
export { default as getUserInfo } from './getUserInfo';
export { default as listUsers } from './listUsers';
export { default as createToken } from './createToken';
export { default as createUser } from './createUser';
export { default as getUmaConfiguration } from './getUmaConfiguration';
export { default as getOpenIdConfiguration } from './getOpenIdConfiguration';
export { default as validateToken } from './validateToken';
