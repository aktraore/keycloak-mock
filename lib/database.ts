import { v4 as uuidv4 } from 'uuid';

import { DuplicateUserError, UserNotFoundError } from './error';

export type MockUserProfileAttributes = {
    [key: string]: [string];
};

export enum MockUserCredentialType {
    PASSWORD = 'password',
    CLIENT_SECRET = 'client_secret',
}

export interface CreateMockUserOptions {
    id?: string;
    createdTimestamp?: number;
    username?: string;
    enabled?: boolean;
    totp?: boolean;
    emailVerified?: boolean;
    firstName?: string;
    lastName?: string;
    email?: string;
    attributes?: MockUserProfileAttributes;
    credentials?: { type?: MockUserCredentialType; value: string }[];
}

export interface UpdateMockUserOptions {
    id: string;
    createdTimestamp?: number;
    username?: string;
    enabled?: boolean;
    totp?: boolean;
    emailVerified?: boolean;
    firstName?: string;
    lastName?: string;
    email?: string;
    attributes?: MockUserProfileAttributes;
    credentials?: { type?: MockUserCredentialType; value: string }[];
}

export interface MockUserProfile {
    id: string;
    createdTimestamp: number;
    username: string;
    enabled: boolean;
    totp: boolean;
    emailVerified: boolean;
    firstName: string;
    lastName: string;
    email: string;
    attributes: MockUserProfileAttributes;
}

export interface MockUser {
    profile: MockUserProfile;
    credentials: { type: MockUserCredentialType; value: string }[];
}

function isNil(value: any): value is null | undefined {
    return value == null;
}

class MockDatabase {
    static SERVICE_USER_EMAIL: string = 'service@keycloak-mock.com';

    users: MockUser[];

    constructor() {
        this.users = [];
    }

    /**
     * Gets all users (except service users).
     */
    allUsers(): MockUser[] {
        return this.users.filter(
            (storedUser) => storedUser.profile.email !== MockDatabase.SERVICE_USER_EMAIL,
        );
    }

    /**
     * Gets all users (except service users).
     */
    filterUsers(filterFunc: (user: MockUser) => boolean): MockUser[] {
        return this.allUsers().filter(filterFunc);
    }

    /**
     * Finds the service account user if any exists.
     */
    findServiceUser(): MockUser | null {
        const user = this.users.find(
            (storedUser) => storedUser.profile.email === MockDatabase.SERVICE_USER_EMAIL,
        );
        return user || null;
    }

    /**
     * Finds an existing user by ID.
     */
    findUserByID(id: string): MockUser | null {
        const user = this.users.find((storedUser) => storedUser.profile.id === id);
        return user || null;
    }

    /**
     * Finds an existing user by username.
     */
    findUserByUsername(username: string): MockUser | null {
        const user = this.users.find((storedUser) => storedUser.profile.username === username);
        return user || null;
    }

    /**
     * Finds an existing user by username.
     */
    findUserByEmail(email: string): MockUser | null {
        const user = this.users.find((storedUser) => storedUser.profile.email === email);
        return user || null;
    }

    /**
     * Deletes a user by ID.
     */
    deleteUserByID(id: string): void {
        this.users = this.users.filter((storedUser) => storedUser.profile.id !== id);
    }

    /**
     * Attempts to match against a user with the specified
     * username and password.
     */
    matchForPasswordGrant(username: string, password: string): MockUser | null {
        const user = this.findUserByUsername(username);
        if (!user) {
            return null;
        }

        const credential = user.credentials.find(
            ({ type }) => type === MockUserCredentialType.PASSWORD,
        );
        if (!credential) {
            return null;
        }

        if (credential.value !== password) {
            return null;
        }

        return user;
    }

    /**
     * Attempts to match against a service account user.
     */
    matchForClientGrant(clientID: string, clientSecret: string): MockUser | null {
        const user = this.findUserByUsername(clientID);
        if (!user) {
            return null;
        }

        const credential = user.credentials.find(
            ({ type }) => type === MockUserCredentialType.CLIENT_SECRET,
        );
        if (!credential) {
            return null;
        }

        if (credential.value !== clientSecret) {
            return null;
        }

        return user;
    }

    /**
     * Deletes all existing users.
     */
    clear(): void {
        this.users = this.users.filter(
            (user) => user.profile.email === MockDatabase.SERVICE_USER_EMAIL,
        );
    }

    /**
     * Creates a new user and returns the profile of the newly created user.
     */
    createUser(options?: CreateMockUserOptions): MockUser {
        const finalizedOptions = options || {};

        const id = finalizedOptions.id || uuidv4();
        const email = finalizedOptions.email || 'henk.jansen@gmail.com';
        const username = finalizedOptions.username || email;

        if (this.findUserByID(id)) {
            throw new DuplicateUserError(`A user with id ${id} already exists`);
        }

        if (this.findUserByUsername(username)) {
            throw new DuplicateUserError(`A user with username ${username} already exists`);
        }

        if (this.findUserByEmail(email)) {
            throw new DuplicateUserError(`A user with email ${email} already exists`);
        }

        const profile: MockUserProfile = {
            id,
            createdTimestamp: finalizedOptions.createdTimestamp || new Date().getTime(),
            username,
            enabled: isNil(finalizedOptions.enabled) ? true : finalizedOptions.enabled,
            totp: isNil(finalizedOptions.totp) ? true : finalizedOptions.totp,
            emailVerified: isNil(finalizedOptions.emailVerified)
                ? true
                : finalizedOptions.emailVerified,
            firstName: finalizedOptions.firstName || 'Henk',
            lastName: finalizedOptions.lastName || 'Jansen',
            email,
            attributes: {
                ...(finalizedOptions.attributes || {}),
            },
        };

        const credentials = finalizedOptions.credentials || [];

        const user: MockUser = {
            profile,
            credentials: credentials.map(({ type, value }) => ({
                type: type || MockUserCredentialType.PASSWORD,
                value,
            })),
        };

        this.users.push(user);

        return user;
    }

    /**
     * Creates a service account that can authenticate using
     * client credentials.
     */
    createServiceUser(clientID: string, clientSecret: string): MockUser {
        if (this.findServiceUser()) {
            throw new Error('There can only be one service account.');
        }

        return this.createUser({
            username: clientID,
            email: MockDatabase.SERVICE_USER_EMAIL,
            enabled: true,
            emailVerified: true,
            credentials: [
                {
                    type: MockUserCredentialType.CLIENT_SECRET,
                    value: clientSecret,
                },
            ],
        });
    }

    /**
     * Updates a user and returns the update profile of the user
     *
     */
    updateUser(options?: UpdateMockUserOptions): MockUser {
        const finalizedOptions = options;

        const id = finalizedOptions?.id;
        const email = finalizedOptions?.email;

        if (!id) {
            throw new UserNotFoundError(`A user with id ${id} doesn't exist`);
        }
        let indexInUser = 0;

        const userToUpdate = this.users.find((storedUser, index) => {
            indexInUser = 0;
            return storedUser.profile.id === id;
        });

        if (!userToUpdate) {
            throw new UserNotFoundError(`A user with id ${id} doesn't exist`);
        }

        const updatedUser: MockUser = {
            profile: {
                ...userToUpdate.profile,
                ...options,
            },
            credentials: userToUpdate.credentials,
        };

        this.users[indexInUser] = updatedUser;
        return updatedUser;
    }
}

export default MockDatabase;
