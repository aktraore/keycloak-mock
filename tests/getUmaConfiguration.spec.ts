import axios from 'axios';

import * as KeycloakMock from '../lib';
import { setupBefore, teardownAfter, getMockInstance } from './util';

describe('getUmaConfiguration', () => {
    beforeAll(setupBefore);
    afterAll(teardownAfter);

    it('works without token', async () => {
        const kmock = getMockInstance();
        const url = kmock.createURL(`/realms/${kmock.params.realm}/.well-known/uma2-configuration`);

        const response = await axios.get(url, { validateStatus: () => true });
        expect(response.status).toBe(200);
        expect(response.data).toMatchSnapshot();
    });

    it('works with token', async () => {
        const kmock = getMockInstance();

        const user = kmock.database.users[0];
        const token = kmock.createBearerToken(user.profile.id);

        const url = kmock.createURL(`/realms/${kmock.params.realm}/.well-known/uma2-configuration`);

        const response = await axios.get(url, {
            headers: { authorization: `Bearer ${token}` },
        });
        expect(response.status).toBe(200);
        expect(response.data).toMatchSnapshot();
    });
});
