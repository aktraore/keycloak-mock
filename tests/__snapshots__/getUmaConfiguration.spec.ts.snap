// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getUmaConfiguration works with token 1`] = `
Object {
  "authorization_endpoint": "http://localhost/auth/realms/myrealm/protocol/openid-connect/auth",
  "end_session_endpoint": "http://localhost/auth/realms/myrealm/protocol/openid-connect/logout",
  "frontchannel_logout_session_supported": true,
  "frontchannel_logout_supported": true,
  "grant_types_supported": Array [
    "authorization_code",
    "implicit",
    "refresh_token",
    "password",
    "client_credentials",
    "urn:ietf:params:oauth:grant-type:device_code",
    "urn:openid:params:grant-type:ciba",
  ],
  "introspection_endpoint": "http://localhost/auth/realms/myrealm/protocol/openid-connect/token/introspect",
  "issuer": "http://localhost/auth/realms/myrealm",
  "jwks_uri": "http://localhost/auth/realms/myrealm/protocol/openid-connect/certs",
  "permission_endpoint": "http://localhost/auth/realms/myrealmz/protection/permission",
  "policy_endpoint": "http://localhost/auth/realms/myrealmz/protection/uma-policy",
  "registration_endpoint": "http://localhost/auth/realms/myrealm/clients-registrations/openid-connect",
  "resource_registration_endpoint": "http://localhost/auth/realms/myrealmz/protection/resource_set",
  "response_modes_supported": Array [
    "query",
    "fragment",
    "form_post",
    "query.jwt",
    "fragment.jwt",
    "form_post.jwt",
    "jwt",
  ],
  "response_types_supported": Array [
    "code",
    "none",
    "id_token",
    "token",
    "id_token token",
    "code id_token",
    "code token",
    "code id_token token",
  ],
  "scopes_supported": Array [
    "openid",
    "profile",
    "microprofile-jwt",
    "offline_access",
    "address",
    "web-origins",
    "phone",
    "roles",
    "email",
  ],
  "token_endpoint": "http://localhost/auth/realms/myrealm/protocol/openid-connect/token",
  "token_endpoint_auth_methods_supported": Array [
    "private_key_jwt",
    "client_secret_basic",
    "client_secret_post",
    "tls_client_auth",
    "client_secret_jwt",
  ],
  "token_endpoint_auth_signing_alg_values_supported": Array [
    "PS384",
    "ES384",
    "RS384",
    "HS256",
    "HS512",
    "ES256",
    "RS256",
    "HS384",
    "ES512",
    "PS256",
    "PS512",
    "RS512",
  ],
}
`;

exports[`getUmaConfiguration works without token 1`] = `
Object {
  "authorization_endpoint": "http://localhost/auth/realms/myrealm/protocol/openid-connect/auth",
  "end_session_endpoint": "http://localhost/auth/realms/myrealm/protocol/openid-connect/logout",
  "frontchannel_logout_session_supported": true,
  "frontchannel_logout_supported": true,
  "grant_types_supported": Array [
    "authorization_code",
    "implicit",
    "refresh_token",
    "password",
    "client_credentials",
    "urn:ietf:params:oauth:grant-type:device_code",
    "urn:openid:params:grant-type:ciba",
  ],
  "introspection_endpoint": "http://localhost/auth/realms/myrealm/protocol/openid-connect/token/introspect",
  "issuer": "http://localhost/auth/realms/myrealm",
  "jwks_uri": "http://localhost/auth/realms/myrealm/protocol/openid-connect/certs",
  "permission_endpoint": "http://localhost/auth/realms/myrealmz/protection/permission",
  "policy_endpoint": "http://localhost/auth/realms/myrealmz/protection/uma-policy",
  "registration_endpoint": "http://localhost/auth/realms/myrealm/clients-registrations/openid-connect",
  "resource_registration_endpoint": "http://localhost/auth/realms/myrealmz/protection/resource_set",
  "response_modes_supported": Array [
    "query",
    "fragment",
    "form_post",
    "query.jwt",
    "fragment.jwt",
    "form_post.jwt",
    "jwt",
  ],
  "response_types_supported": Array [
    "code",
    "none",
    "id_token",
    "token",
    "id_token token",
    "code id_token",
    "code token",
    "code id_token token",
  ],
  "scopes_supported": Array [
    "openid",
    "profile",
    "microprofile-jwt",
    "offline_access",
    "address",
    "web-origins",
    "phone",
    "roles",
    "email",
  ],
  "token_endpoint": "http://localhost/auth/realms/myrealm/protocol/openid-connect/token",
  "token_endpoint_auth_methods_supported": Array [
    "private_key_jwt",
    "client_secret_basic",
    "client_secret_post",
    "tls_client_auth",
    "client_secret_jwt",
  ],
  "token_endpoint_auth_signing_alg_values_supported": Array [
    "PS384",
    "ES384",
    "RS384",
    "HS256",
    "HS512",
    "ES256",
    "RS256",
    "HS384",
    "ES512",
    "PS256",
    "PS512",
    "RS512",
  ],
}
`;
